documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -111,8 +131,49 @@ Threading rules for the reference implementation (if any) will be explicitly doc
+
+```cs
+public interface IActor
+{
+    void HandleDeviceConnectionChanged(ConnectionEvent<IInputDevice> @event);
+}
+```
+
+The base `IActor` is simple, handling only device connections and device disconnections.
+
+`HandleDeviceConnectionChanged` must be called with a device and a value of `true` if the `ConnectionEvent.Device` has just been added to the `IInputBackend.Devices` list.
+
+`HandleDeviceConnectionChanged` must be called with a device and a value of `false` if the `ConnectionEvent.Device` has just been removed from the `IInputBackend.Devices` list.
+
+All actor methods are called in the order that the state changes happened in the underlying backend. For example, a "double click" would be mouse down, mouse up, mouse down, and mouse up again. Even if those events happened entirely between two update calls (thus resulting in a net-zero change of state, given the button was released at the last update and is still released at this update), the input backend is expected to queue up the events so they may be delivered to the actor in the order in which they occurred when `Update` is called **where possible**.
+
+**INFORMATIVE TEXT:** Some backends, such as OpenXR, may only operate on a state query basis, and obviously we only query the state when `Update` is called. This is the reason why it's only a "where possible" requirement, because if we can't actually get the events in the order they happened from the underlying backend (like we can for GLFW, for example) then it's impossible for us to convey this up to the developer. However, regardless of backend-specifics, a `State` update must always be paired with an actor method call.
+
+The `IActor` passed into `Update` may implement multiple other actor interfaces (as defined below), and if the actor implements an extra interface (such as `IMouseActor` defined below) that would allow the backend to forward more events to the actor, the backend must do so via type checking. That is, if `actor` is an instance of `IMouseActor`, any mouse events are delivered to that actor. But if `actor` does not implement `IMouseActor`, no mouse events will be delivered.
+
+Note that during the `Update` call, a backend must only update the device's state in the order that the events are delivered. For example when `IInputBackend.Update` is called:
+1. The backend has a queued "mouse down" event.
+2. The backend updates the `State` of the relevant `IMouse` for that button press.
+3. The backend calls `HandleButtonDown` on the `IMouseActor` (if applicable).
+4. The backend has a queued "mouse up" event.
+5. The backend updates the `State` of the relevant `IMouse` for that button release.
+6. The backend calls `HandleButtonUp` on the `IMouseActor` (if applicable).
```

Comment:
As of 2.X, input events can occur in between updates (I just tried mouse events with an update rate of 1 per second) which pretty much ensures as good as GLFW can that on button callbacks you get the position of the mouse where the event occur, so this is important if we want if we want to be able to keep that ability of processing the button event at the position it occurred, not at the position the mouse was when the event was finally handled

Reply:
If I understand correctly you're asking "can we get the position of the down/up?" which is resolved by the move and down/up events being sent in order i.e. you'll move your cursor then you'll get a down then an up. This is just an example for just the down/up events - the events will continue to function as in 2.X today and the state will reflect the order in which the events are sent.

Reply:
Yes (I commented that just to make sure we're on the same page)

Reply:
Awesome, thanks :D